### PR TITLE
Port Site to Python3

### PR DESCRIPTION
**Pull request by kiritofeng**

### CODE REVIEW

File: event_socket_server/helpers.py [hunk 1]
```diff
@@ -38,14 +39,14 @@ def send(self, data, callback=None):
 
 class ZlibPacketHandler(SizedPacketHandler):
     def _format_send(self, data):
-        return data.encode('zlib')
+        return codecs.encode(data.encode('utf-8'), 'zlib')
```
Comment (quantum5) [inline event_socket_server/helpers.py]: Can we just use `zlib.compress`? `codecs` is a bit of a hack...

File: event_socket_server/helpers.py [hunk 2]
```diff
@@ -38,14 +39,14 @@ def send(self, data, callback=None):
...
 
     def packet(self, data):
         raise NotImplementedError()
 
     def _packet(self, data):
         try:
-            self.packet(data.decode('zlib'))
+            self.packet(codecs.decode(data, 'zlib').decode('utf-8'))
```
Comment (quantum5) [inline event_socket_server/helpers.py]: Can we just use `zlib.decompress`? `codecs` is a bit of a hack...

File: event_socket_server/test_client.py [hunk 3]
```diff
@@ -16,14 +17,14 @@ def open_connection():
 
 
 def zlibify(data):
-    data = data.encode('zlib')
+    data = codecs.encode(data.encode('utf-8'), 'zlib')
```
Comment (quantum5) [inline event_socket_server/test_client.py]: Can we just use `zlib.compress`? `codecs` is a bit of a hack...

File: event_socket_server/test_client.py [hunk 4]
```diff
@@ -16,14 +17,14 @@ def open_connection():
...
     return size_pack.pack(len(data)) + data
 
 
 def dezlibify(data, skip_head=True):
     if skip_head:
         data = data[size_pack.size:]
-    return data.decode('zlib')
+    return codecs.decode(data.decode('utf-8'), 'zlib')
```
Comment (quantum5) [inline event_socket_server/test_client.py]: Can we just use `zlib.decompress`? `codecs` is a bit of a hack...

File: judge/admin/submission.py [hunk 5]
```diff
@@ -202,11 +202,11 @@ def execution_time(self, obj):
...
         memory = obj.memory
         if memory is None:
-            return ugettext('None')
+            return gettext('None')
         if memory < 1000:
-            return ugettext('%d KB') % memory
+            return gettext('%d KB') % memory
         else:
-            return ugettext('%.2f MB') % (memory / 1024.)
+            return gettext('%.2f MB') % (memory / 1024.)
```
Comment (quantum5) [inline judge/admin/submission.py]: No need for the `.` in the end since Python 3.

File: judge/jinja2/gravatar.py [hunk 6]
```diff
@@ -17,7 +17,8 @@ def gravatar(email, size=80, default=None):
     elif isinstance(email, AbstractUser):
         email = email.email
 
-    gravatar_url = '//www.gravatar.com/avatar/' + hashlib.md5(utf8bytes(email.strip().lower())).hexdigest() + '?'
+    gravatar_url = '//www.gravatar.com/avatar/' + hashlib.md5(utf8bytes(email.strip().lower().encode('utf-8'))) \
```
Comment (quantum5) [inline judge/jinja2/gravatar.py]: Wait... why `.encode('utf-8')` if we already have `utf8bytes`?

File: judge/judgeapi.py [hunk 7]
```diff
@@ -16,14 +17,14 @@ def judge_request(packet, reply=True):
                                     settings.BRIDGED_DJANGO_ADDRESS[0])
 
     output = json.dumps(packet, separators=(',', ':'))
-    output = output.encode('zlib')
-    writer = sock.makefile('w', 0)
+    output = codesc.encode(output.encode(), 'zlib')
```
Comment (quantum5) [inline judge/judgeapi.py]: `codesc.encode`.... :wtf: What happened to... `zlib.compress`?

File: judge/judgeapi.py [hunk 8]
```diff
@@ -34,8 +35,8 @@ def judge_request(packet, reply=True):
         reader.close()
         sock.close()
 
-        input = input.decode('zlib')
-        result = json.loads(input)
+        input = codecs.decode(input, 'zlib')
```
Comment (quantum5) [inline judge/judgeapi.py]: Use `zlib.decompress` and inline `input`.

File: judge/models/choices.py [hunk 9]
```diff
@@ -14,7 +14,7 @@ def make_timezones():
             area, loc = 'Other', tz
         if not loc.startswith('GMT'):
             data[area].append((tz, loc))
-    data = data.items()
+    data = list(data.items())
```
Comment (quantum5) [inline judge/models/choices.py]: Consider `return sorted(data.items(), key=itemgetter(0))`

File: judge/models/profile.py [hunk 10]
```diff
@@ -120,7 +120,7 @@ def organization(self):
     def username(self):
         return self.user.username
 
-    def calculate_points(self, table=(lambda x: [pow(x, i) for i in xrange(100)])(getattr(settings, 'PP_STEP', 0.95))):
+    def calculate_points(self, table=(lambda x: [pow(x, i) for i in range(100)])(getattr(settings, 'PP_STEP', 0.95))):
```
Comment (quantum5) [inline judge/models/profile.py]: Why is this not refactored into something sane???

File: judge/models/runtime.py [hunk 11]
```diff
@@ -55,7 +55,7 @@ def runtime_versions(self):
             runtimes[id].add(runtime.version)
 
         lang_versions = []
-        for id, version_list in runtimes.iteritems():
+        for id, version_list in runtimes.items():
             lang_versions.append((id, list(sorted(version_list, key=lambda a: tuple(map(int, a.split('.')))))))
```
Comment (quantum5) [inline judge/models/runtime.py]: Did someone just do... `list(sorted(...))`?

File: judge/utils/caniuse.py [hunk 12]
```diff
@@ -1,10 +1,10 @@
 import json
-import urllib2
+import urllib.request
 from contextlib import closing
 
 from ua_parser import user_agent_parser
 
-with closing(urllib2.urlopen('https://raw.githubusercontent.com/Fyrd/caniuse/master/data.json')) as f:
+with closing(urllib.request.urlopen('https://raw.githubusercontent.com/Fyrd/caniuse/master/data.json')) as f:
```
Comment (quantum5) [inline judge/utils/caniuse.py]: We already have a `requests` dependency... 🤔 though this is probably for later.

File: judge/utils/raw_sql.py [hunk 13]
```diff
@@ -52,7 +58,56 @@ def join(self, join, reuse=None):
...
+        pass
+    assert (type(obj) == type(old_query))
+    return obj
+
+
 def use_straight_join(queryset):
     if connections[queryset.db].vendor != 'mysql':
         return
-    queryset.query = queryset.query.clone(straight_join_cache[type(queryset.query)])
+    queryset.query = clone_to_new(queryset.query)
```
Comment (quantum5) [inline judge/utils/raw_sql.py]: Revert this. This is django 2.0 magic that I'll deal with later.

**Comment by kiritofeng** [comment]
Things tested:
1. Submitting to a problem.
2. Creating a contest.
3. Joining and submitting to a contest.
4. Creating, responding to, and closing a ticket.
5. Commenting.

File: judge/models/problem.py [hunk 14]
```diff
@@ -247,21 +249,24 @@ def clarifications(self):
...
+                                                          user__is_unlisted=False).count() / submissions if submissions else 0
         self.save()
 
     update_stats.alters_data = True
 
     def _get_limits(self, key):
         global_limit = getattr(self, key)
         limits = {limit['language_id']: (limit['language__name'], limit[key])
-                  for limit in self.language_limits.values('language_id', 'language__name', key) if limit[key] != global_limit}
+                  for limit in self.language_limits.values('language_id', 'language__name', key) if
```
Comment (quantum5) [inline judge/models/problem.py]: If should go on the next line...

File: judge/utils/caniuse.py [hunk 15]
```diff
@@ -1,11 +1,13 @@
 import json
-import urllib2
+import urllib.request
```
Comment (quantum5) [inline judge/utils/caniuse.py]: Can we get rid of this?

File: judge/views/widgets.py [hunk 16]
```diff
@@ -1,12 +1,15 @@
 import json
-import urllib2
+import urllib.request
```
Comment (quantum5) [inline judge/views/widgets.py]: Do we still need this?

File: judge/views/mailgun.py [hunk 17]
```diff
@@ -26,7 +28,7 @@ def post(self, request, *args, **kwargs):
 
             logger.debug('Received request: %s', params)
 
-            if signature != hmac.new(key=settings.MAILGUN_ACCESS_KEY, msg='%s%s' % (timestamp, token),
+            if signature != hmac.new(utf8bytes(key=settings.MAILGUN_ACCESS_KEY), msg=b'%s%s' % (timestamp, token),
```
Comment (quantum5) [inline judge/views/mailgun.py]: Uhhh... No. You can't just `b'%s%s'` when the arguments are strings. You have to format it normally and then `utf8bytes` it.

File: judge/views/status.py [hunk 18]
```diff
@@ -88,23 +88,23 @@ def version_matrix(request):
                     size[i] += 1
                     size[j] = 0
 
-        rep = max(xrange(len(data)), key=size.__getitem__)
+        rep = list(max(range(len(data)), key=size.__getitem__))
```
Comment (quantum5) [inline judge/views/status.py]: `list(max(...))`, really?

File: judge/views/blog.py [hunk 19]
```diff
@@ -91,7 +92,7 @@ def get_context_data(self, **kwargs):
         # Superusers better be staffs, not the spell-casting kind either.
         if self.request.user.is_staff:
             tickets = (Ticket.objects.order_by('-id').filter(is_open=True).prefetch_related('linked_item')
-                             .select_related('user__user'))
+                       .select_related('user__user'))
```
Comment (Xyene) [inline judge/views/blog.py]: Indent.

File: judge/views/blog.py [hunk 20]
```diff
@@ -30,15 +30,16 @@ def get_paginator(self, queryset, per_page, orphans=0,
...
 
     def get_context_data(self, **kwargs):
         context = super(PostList, self).get_context_data(**kwargs)
         context['title'] = self.title or _('Page %d of Posts') % context['page_obj'].number
         context['first_page_href'] = reverse('home')
         context['page_prefix'] = reverse('blog_post_list')
         context['comments'] = Comment.most_recent(self.request.user, 10)
-        context['new_problems'] = Problem.objects.filter(is_public=True, is_organization_private=False).order_by('-date', '-id')[:7]
+        context['new_problems'] = Problem.objects.filter(is_public=True, is_organization_private=False) \
+                                      .order_by('-date', '-id')[:7]
```
Comment (Xyene) [inline judge/views/blog.py]: Indent.

File: judge/views/api/api_v2.py [hunk 21]
```diff
@@ -115,7 +116,8 @@ def api_v2_user_info(request):
         'points': profile.points,
         'solved': solved_problems,
         'attempted': attempted_problems,
-        'authored': list(Problem.objects.filter(is_public=True, is_organization_private=False, authors=profile).values_list('code', flat=True))
+        'authored': list(Problem.objects.filter(is_public=True, is_organization_private=False, authors=profile)
+                         .values_list('code', flat=True))
```
Comment (Xyene) [inline judge/views/api/api_v2.py]: Indent.

File: judge/views/register.py [hunk 22]
```diff
@@ -42,14 +42,14 @@ class CustomRegistrationForm(RegistrationForm):
...
+            raise forms.ValidationError(gettext('The email address "%s" is already taken. Only one registration '
+                                                 'is allowed per address.') % self.cleaned_data['email'])
         if '@' in self.cleaned_data['email']:
             domain = self.cleaned_data['email'].split('@')[-1].lower()
             if (domain in getattr(settings, 'BAD_MAIL_PROVIDERS', ())
                     or any(regex.match(domain) for regex in bad_mail_regex)):
-                raise forms.ValidationError(ugettext(u'Your email provider is not allowed due to history of abuse. '
-                                                     u'Please use a reputable email provider.'))
+                raise forms.ValidationError(gettext('Your email provider is not allowed due to history of abuse. '
+                                                     'Please use a reputable email provider.'))
```
Comment (Xyene) [inline judge/views/register.py]: One space too much indent.

File: judge/views/register.py [hunk 23]
```diff
@@ -42,14 +42,14 @@ class CustomRegistrationForm(RegistrationForm):
 
     def clean_email(self):
         if User.objects.filter(email=self.cleaned_data['email']).exists():
-            raise forms.ValidationError(ugettext(u'The email address "%s" is already taken. Only one registration '
-                                                 u'is allowed per address.') % self.cleaned_data['email'])
+            raise forms.ValidationError(gettext('The email address "%s" is already taken. Only one registration '
+                                                 'is allowed per address.') % self.cleaned_data['email'])
```
Comment (Xyene) [inline judge/views/register.py]: One space too much indent.

File: judge/models/problem.py [hunk 24]
```diff
@@ -106,8 +106,10 @@ class Problem(models.Model):
                                          'These users will be able to view a private problem, but not edit it.'))
     types = models.ManyToManyField(ProblemType, verbose_name=_('problem types'))
     group = models.ForeignKey(ProblemGroup, verbose_name=_('problem group'))
-    time_limit = models.FloatField(verbose_name=_('time limit'), help_text=_('The time limit for this problem, in seconds. Fractional seconds (e.g. 1.5) are supported.'))
-    memory_limit = models.IntegerField(verbose_name=_('memory limit'), help_text=_('The memory limit for this problem, in kilobytes (e.g. 64mb = 65536 kilobytes).'))
+    time_limit = models.FloatField(verbose_name=_('time limit'), help_text=_(
+        'The time limit for this problem, in seconds. Fractional seconds (e.g. 1.5) are supported.'))
```
Comment (Xyene) [inline judge/models/problem.py]: Gross indent.

**Pull request merged**